### DISCUSSION TITLE
What replication to use

### DISCUSSION THREAD

**Discussion by xinix00** [ravendb/ravendb#12596] [General]
1. Customer wants 2 servers, so a Cluster might not be the way to go. Also, since data must be replicated to a SQL Server, our ID's must be INT, when using a Cluster, it MIGHT get users/123-A and users/123-B, which doesn't translate well to ID 123.
2. In RavenDB 3.0 it was possible to create Replication both ways, don't know if this is currently possible or that it's only one way?
3. Use the Hub/Sink method, which allows both way communication.

Client wants to have 1 main server, and one backup in case of loss of the first one.

P.s. in order to use cluster, I could also rewrite the ID generator to give ID's like: users/123 1 and users/123 2 (where 1 = cluster A, 2 is cluster B).

Maybe you could also enlighten me how big the chance is that a collision occurs between the 2 clusters (so get users/123-A and users/123-B), or maybe it's possible to use 1 node as WRITE only and both for READ. I found an old blog post of 2019 stating that 2 node clusters are possible, but that you have to make 1 server a Watcher, not Member.

**Comment by ayende** [comment]
* You control your ids, you don't _have_ to use the `-A` suffix, although we recommend it to avoid conflicts.
* You can setup external replication between nodes, similar to how it worked in 3.5
* If you have just two of them, external replication is easier.

If you have a backup server, how do you switch to it? If this is a cluster, it will happen automatically.
If those are separate instances connected via external replication, you'll have to update the URL.

The issue of collision is that you may have the same action happen on both nodes to reserve a range at the same time.

In a two node cluster, we recommend having the main node as a member and the other as a watcher, yes.

**Comment by xinix00** [comment]
Since cluster might be a way to go, how high is the chance of a collision, let's say we got 2 users, on 2 PCs running, is it possible that the first "client1" will get HiLo with max 32, and "client2" which is connected to node2, also gets HiLo with max 32 (so they will both generate the same ID's, when I disable the suffix -A/-B?

If that's true, a client might have to press 16 or 32 times on save (don't know how high the hilo bump is) to fix the "issue" of conflicts.

Is this chance the same on double replication? Since we used that for a long time but never had an issue with conflicts (well ok, rarely)?

Or is it maybe possible to set a server to WRITES/READS only and the other server to READS so that conflicts can't happen... since only one server is touching the HiLo.

**Comment by ayende** [comment]
There is a chance for that, yes. In terms of likelihood, not much. You need to have a disconnect in the network, two clients that reach different nodes at the same time, etc. 

You also need multiple clients in this scenario, mind.

**Comment by ayende** [comment]
By default, we have a single writer active, the second one is only active is there is a failur.

**Comment by xinix00** [comment]
There are multiple clients, I don't know how RavenDB will place them, but if it normally (when both nodes are up) the HiLo won't fail, than that's good enough for me. But I don't know if different clients will read from different nodes, and are able to get a HiLo which is already changed on another server to a higher number.

So Client 1 -> Node A
Client 2 -> Node B

Or will Client 1 and Client 2 normally always connect to 1 node to get the HiLo from. Else it looks like there is a small chance that the HiLo is changed on Node A, not yet replicated on Node B, and client 2 therefor get's an older HiLo and starts generating duplicates / conflicts.

**Reply by ayende** [reply]
Both clients will use the same node as long as there aren't any failures.
